src/chain/shared_state.rs

```diff
@@ -537,8 +543,111 @@ impl SectionProofChain {
         self.blocks.len()
     }
 
+    pub fn last_prefix_version(&self) -> (&Prefix<XorName>, u64) {
+        self.blocks
+            .last()
+            .map(|block| (block.prefix(), block.version()))
+            .unwrap_or((&self.prefix, self.version))
+    }
+
+    #[cfg(all(test, feature = "mock"))]
+    pub fn all_prefix_version(&self) -> impl DoubleEndedIterator<Item = (&Prefix<XorName>, u64)> {
+        iter::once((&self.prefix, self.version)).chain(
+            self.blocks
+                .iter()
+                .map(|block| (block.prefix(), block.version())),
+        )
+    }
+
+    pub fn last_new_public_key_info(&self) -> Option<&SectionKeyInfo> {
+        self.blocks.last().map(|block| block.key_info())
+    }
+
+    fn last_trusted_key_info<'a>(
+        &'a self,
+        last_trusted: &'a SectionKeyInfo,
+    ) -> Option<&'a SectionKeyInfo> {
+        let block_offset = last_trusted.version().saturating_sub(self.version) as usize;
+
+        if block_offset == 0 {
+            if last_trusted.version() != self.version || last_trusted.prefix() != &self.prefix {
+                return None;
+            }
+        } else if let Some(block) = self.blocks.get(block_offset - 1) {
+            if block.key_info() != last_trusted {
+                return None;
+            }
+        } else {
+            // Root of trust not found
+            return None;
+        }
+
+        let mut current = last_trusted;
```

Comment:
yes, because block[3] is version 6. (i.e if self.version is 2, then block[0] is version 3, block[1] is version 4, block[2] is version 5, block[3] is version 6.